File: src/amazonqGumby/telemetry/codeTransformTelemetryState.ts
/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { randomUUID } from 'crypto'

interface ICodeTransformerTelemetryState {
    sessionId: string
    sessionStartTime: number
}

class CodeTransformerTelemetryState {
    private static instance: CodeTransformerTelemetryState
    mainState: ICodeTransformerTelemetryState

    private constructor() {
        this.mainState = {
            sessionId: randomUUID(),
            sessionStartTime: Date.now(),
        }
    }

    public static getInstance(): CodeTransformerTelemetryState {
        if (!CodeTransformerTelemetryState.instance) {
            CodeTransformerTelemetryState.instance = new CodeTransformerTelemetryState()
        }

        return CodeTransformerTelemetryState.instance
    }

    public getSessionId = () => this.mainState.sessionId
    public getStartTime = () => this.mainState.sessionStartTime

    public setSessionId = () => {
        this.mainState.sessionId = randomUUID()
    }
    public setStartTime = () => {
        this.mainState.sessionStartTime = Date.now()
    }
}

export const codeTransformTelemetryState = CodeTransformerTelemetryState.getInstance()

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/state/CodeTransformTelemetryState.kt
// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.services.codemodernizer.state

import com.intellij.openapi.components.BaseState
import com.intellij.util.xmlb.annotations.Property
import java.time.Instant
import java.util.UUID

/**
 * The persistent state for the [CodeModernizerTelemetryManager] as such should only be accessed by [CodeModernizerTelemetryManager].
 */
class CodeTransformTelemetryState {
    private val mainState = CodeModernizerTelemetryStateBase()

    fun getSessionId() = mainState.sessionId
    fun setSessionId() {
        mainState.sessionId = UUID.randomUUID().toString()
    }

    fun getStartTime() = mainState.sessionStartTime
    fun setStartTime() {
        mainState.sessionStartTime = Instant.now()
    }

    // Companion object to hold the singleton instance
    companion object {
        // Lazy initialization of the singleton instance
        val instance: CodeTransformTelemetryState by lazy { CodeTransformTelemetryState() }
    }
}

class CodeModernizerTelemetryStateBase : BaseState() {
    @get:Property
    var sessionId: String = UUID.randomUUID().toString()

    @get:Property
    var sessionStartTime: Instant = Instant.now()
}
